src/patch.c

```diff
@@ -180,30 +180,109 @@ Patch_create_from(PyObject *self, PyObject *args, PyObject *kwds)
   return wrap_patch(patch, oldblob, newblob);
 }
 
+PyObject *
+decode(Patch *patch, char *encoding, char *errors)
+{
+    git_buf buf = {NULL};
+    int err;
+    PyObject *ret;
+
+    assert(patch->patch);
+    err = git_patch_to_buf(&buf, patch->patch);
+    if (err < 0)
+        return Error_set(err);
+
+    ret = PyUnicode_Decode(buf.ptr, buf.size, encoding, errors);
+    git_buf_free(&buf);
+    return ret;
+}
 
 PyDoc_STRVAR(Patch_patch__doc__,
-    "Patch diff string. Can be None in some cases, such as empty commits.");
+    "Patch diff string (deprecated -- use Patch.decode() instead).\n"
+    "Can be None in some cases, such as empty commits. "
+    "Note that this decodes the content to unicode assuming UTF-8 encoding. "
+    "For non-UTF-8 content that can lead be a lossy, non-reversible process. "
+    "To access the raw, un-decoded patch, use `str(patch)` (Python 2), or "
+    "`bytes(patch)` (Python 3).");
 
 PyObject *
 Patch_patch__get__(Patch *self)
+{
+    PyErr_WarnEx(PyExc_DeprecationWarning, "`Patch.patch` assumes UTF-8 encoding and can have unexpected results on "
+                                           "other encodings. If decoded text is needed, use `Patch.decode()` "
+                                           "instead. Otherwise use `bytes(Patch)`.", 1);
+    return decode(self, "utf-8", "replace");
```

Comment:
These hardwired parameters are equivalent to the original logic of `to_unicode(buf.ptr, NULL, NULL)`.